File: qeli-ios/QeliCore/Model/Protection.swift
import Foundation

// `PushedFacts` used to live here. It moved to PushedFacts.swift: `TunnelSnapshot` carries
// it, and the widget target compiles TunnelSnapshot from an explicit file list, so keeping it
// next to `ProtectionSummary` — which is built from a `VPNConfig` — would pull the config
// model into an extension that needs none of it.

/// How much of the device's traffic the tunnel carries.
///
/// Structurally identical to the Android `ProtectionScope` so the two cards stay comparable
/// case for case, but **iOS only ever produces `.all` and `.splitRoutes`** — see
/// `.onlySelected` below. The unreachable cases are kept for that parity (and the UI keeps
/// strings for them), NOT because this platform can reach them.
enum ProtectionScope: Equatable, Sendable {
    /// Every app, every route.
    case all
    /// Only the apps the user picked (`apps_mode = include`).
    ///
    /// **Never produced on iOS.** Per-app rules need `NEAppRule`, which needs an MDM-managed
    /// configuration, so the selection is not in force and reporting it as the scope would
    /// confirm a restriction that does not exist. A profile carrying `apps_mode` gets
    /// `ProtectionWarning.perAppNotApplied` instead. Do not wire this back up without a
    /// managed configuration to back it. (Audit 2026-08-02, §7.)
    case onlySelected
    /// Every app except the ones the user picked (`apps_mode = exclude`).
    /// **Never produced on iOS**, for the same reason as `.onlySelected`.
    case allExcept
    /// Split tunnel: only the configured/pushed routes go through the VPN.
    case splitRoutes
}

/// Something that narrows what the tunnel protects, worth telling the user about.
enum ProtectionWarning: Equatable, Sendable {
    /// `allow_lan` — RFC1918 is carved out, so LAN traffic is not in the tunnel.
    case lanOutside
    /// `allow_ipv4_leak` — native IPv4 may bypass an IPv6-only full tunnel.
    case ipv4Outside
    /// `allow_ipv6_leak` — native IPv6 may bypass an IPv4-only full tunnel.
    case ipv6Outside
    /// Explicit `exclude` routes.
    case excludedRoutes
    /// No pinned server key: the first connection trusts whoever answers (TOFU).
    case noPinnedKey
    /// The profile carries `apps_mode`, but iOS cannot apply per-app rules without MDM
    /// (`NEAppRule` needs a managed configuration), so EVERY app goes through the tunnel
    /// regardless of the selection. (Audit 2026-08-02, §7.)
    case perAppNotApplied
}

extension VPNConfig {
    /// Release metadata resolves through DNS and may use either address family. Only an
    /// unqualified full capture can promise that an app-owned request enters this tunnel.
    func hasPrivateUpdatePath(globalAllowLAN: Bool = false) -> Bool {
        isFullTunnel
            && !allowIPv4Leak
            && !allowIPv6Leak
            && !allowLAN
            && !globalAllowLAN
            && excludeRoutes.isEmpty
    }
}

/// What a profile actually protects, derived from the profile alone.
///
/// Mirror of the Android `ProtectionSummary`, decision for decision — the two cards must
/// never disagree about the same profile.
///
/// This backs a card that makes SECURITY CLAIMS, so the rule is: state only what the config
/// guarantees, and give anything that narrows the tunnel its own line rather than folding it
/// into a reassuring headline. A card that says "all traffic is protected" when it isn't is
/// worse than no card at all.
///
/// Deliberately a pure function of `VPNConfig` with enum outputs — no view code, no
/// localized text — so the decisions are testable and the wording stays localizable.
/// Runtime facts the profile cannot know (which resolver the server actually pushed, the
/// negotiated MTU) are NOT guessed here; they arrive with the tunnel snapshot.
struct ProtectionSummary: Equatable, Sendable {
    let scope: ProtectionScope
    /// Size of the per-app selection; meaningful for `onlySelected` / `allExcept`.
    let appCount: Int
    let excludedRouteCount: Int
    /// X25519 + ML-KEM-768. True for every wire mode except `plain`.
    let postQuantum: Bool
    let dnsThroughTunnel: Bool
    let keyPinned: Bool
    let warnings: [ProtectionWarning]

    private init(
        scope: ProtectionScope,
        appCount: Int,
        excludedRouteCount: Int,
        postQuantum: Bool,
        dnsThroughTunnel: Bool,
        keyPinned: Bool,
        warnings: [ProtectionWarning]
    ) {
        self.scope = scope
        self.appCount = appCount
        self.excludedRouteCount = excludedRouteCount
        self.postQuantum = postQuantum
        self.dnsThroughTunnel = dnsThroughTunnel
        self.keyPinned = keyPinned
        self.warnings = warnings
    }

    /// True only when nothing narrows what the tunnel carries — the one condition under
    /// which the UI may claim "all traffic is protected".
    ///
    /// `noPinnedKey` is excluded on purpose: pinning decides WHO the client is willing to
    /// talk to, not HOW MUCH traffic it carries. It still gets its own warning line.
    /// `perAppNotApplied` is excluded for the opposite reason to `noPinnedKey`: an unapplied
    /// per-app selection does not narrow what the tunnel carries, it WIDENS it — every app
    /// goes through the VPN even though only some were picked. Counting it as a narrowing
    /// would make the card claim less protection than there is. (Audit 2026-08-02, §7.)
    var carriesEverything: Bool {
        scope == .all && warnings.allSatisfy { $0 == .noPinnedKey || $0 == .perAppNotApplied }
    }

    /// - Parameter globalAllowLAN: the app-wide "Allow local network access" setting.
    ///   `TunnelManager` and `QeliNativeTunnelEngine` both carve the private ranges out of a
    ///   full tunnel on `config.allowLAN || settings.allowLAN`, so the card reads the same pair.
    ///   It used to read only the profile field, and with the app-wide switch on it announced
    ///   "all traffic is protected" while RFC1918, link-local and multicast went past the
    ///   VPN — the card erring in the UNSAFE direction, which is the one thing it may never
    ///   do. Mirrors the same fix on Android. (Audit 2026-08-02, §13.)
    init(config: VPNConfig, globalAllowLAN: Bool = false) {
        // `apps_mode` is REPORTED, not applied, on this platform.
        //
        // Consumer iOS cannot install per-app rules: `NEAppRule` requires a managed
        // configuration, and `VPNConfig` says so itself. So a profile with
        // `apps_mode = include` does NOT leave the unselected apps outside — every app goes
        // through the tunnel. Mapping the mode straight onto the scope made the card confirm a
        // restriction that is not in force: the user reads "only the selected apps are
        // protected", arranges their traffic around that belief, and the truth is the
        // opposite. The scope now follows the ROUTES, which is what this platform actually
        // enforces, and the unapplied selection gets its own warning line instead of a
        // headline. (Audit 2026-08-02, §7.)
        let mode = config.appsMode.lowercased()
        let perAppRequested = mode == "include" || mode == "exclude"
        scope = config.isFullTunnel ? .all : .splitRoutes
        appCount = config.apps.count
        excludedRouteCount = config.excludeRoutes.count
        // Every mode runs the hybrid PQ ClientHello except `plain`, which uses a raw X25519
        // exchange. obfs and reality-tls are transport wrappers around the SAME PQ
        // handshake, so they count as post-quantum.
        postQuantum = config.wireMode.lowercased() != "plain"
        // Explicit resolvers are reached through the tunnel; a full tunnel captures DNS
        // regardless. Anything narrower cannot be claimed from the profile alone.
        dnsThroughTunnel = !config.dnsServers.isEmpty || config.isFullTunnel
        keyPinned = !(config.serverPublicKeyHex ?? "").isEmpty

        var found: [ProtectionWarning] = []
        // The compact strip renders only the first warning, so broad missing-family egress
        // must outrank narrower LAN and route exceptions.
        if config.allowIPv4Leak { found.append(.ipv4Outside) }
        if config.allowIPv6Leak { found.append(.ipv6Outside) }
        if !config.excludeRoutes.isEmpty { found.append(.excludedRoutes) }
        if config.isFullTunnel && (config.allowLAN || globalAllowLAN) {
            found.append(.lanOutside)
        }
        if (config.serverPublicKeyHex ?? "").isEmpty { found.append(.noPinnedKey) }
        // Deliberately a warning and NOT part of `carriesEverything`: an unapplied per-app
        // selection does not narrow the tunnel — it widens it beyond what the user asked for.
        // The card must say so without claiming the tunnel protects less than it does.
        if perAppRequested { found.append(.perAppNotApplied) }
        warnings = found
    }
}

extension LiveConnectionProperties {
    /// Build the persisted UI projection from the immutable PacketTunnel configuration.
    init(config: VPNConfig, globalAllowLAN: Bool) {
        let summary = ProtectionSummary(config: config, globalAllowLAN: globalAllowLAN)
        let liveScope: LiveProtectionScope
        switch summary.scope {
        case .all: liveScope = .all
        case .onlySelected: liveScope = .onlySelected
        case .allExcept: liveScope = .allExcept
        case .splitRoutes: liveScope = .splitRoutes
        }
        let liveWarnings = summary.warnings.map { warning -> LiveProtectionWarning in
            switch warning {
            case .lanOutside: return .lanOutside
            case .ipv4Outside: return .ipv4Outside
            case .ipv6Outside: return .ipv6Outside
            case .excludedRoutes: return .excludedRoutes
            case .noPinnedKey: return .noPinnedKey
            case .perAppNotApplied: return .perAppNotApplied
            }
        }
        self.init(
            serverAddress: config.serverAddress,
            port: config.port,
            wireMode: config.wireMode,
            protocolName: config.protocolName,
            quicEnabled: config.quicEnabled,
            configuredMTU: config.mtu,
            reconnectEnabled: config.reconnectEnabled,
            scope: liveScope,
            appCount: summary.appCount,
            excludedRouteCount: summary.excludedRouteCount,
            postQuantum: summary.postQuantum,
            dnsThroughTunnel: summary.dnsThroughTunnel,
            keyPinned: summary.keyPinned,
            warnings: liveWarnings
        )
    }
}

extension ProtectionSummary {
    /// Rehydrate the display model without consulting a profile edited after connect.
    init(live: LiveConnectionProperties) {
        let protectionScope: ProtectionScope
        switch live.scope {
        case .all: protectionScope = .all
        case .onlySelected: protectionScope = .onlySelected
        case .allExcept: protectionScope = .allExcept
        case .splitRoutes: protectionScope = .splitRoutes
        }
        let protectionWarnings = live.warnings.map { warning -> ProtectionWarning in
            switch warning {
            case .lanOutside: return .lanOutside
            case .ipv4Outside: return .ipv4Outside
            case .ipv6Outside: return .ipv6Outside
            case .excludedRoutes: return .excludedRoutes
            case .noPinnedKey: return .noPinnedKey
            case .perAppNotApplied: return .perAppNotApplied
            }
        }
        self.init(
            scope: protectionScope,
            appCount: live.appCount,
            excludedRouteCount: live.excludedRouteCount,
            postQuantum: live.postQuantum,
            dnsThroughTunnel: live.dnsThroughTunnel,
            keyPinned: live.keyPinned,
            warnings: protectionWarnings
        )
    }
}

File: qeli-android/app/src/main/kotlin/com/qeli/model/Protection.kt
package com.qeli.model

/**
 * Negotiated, display-safe runtime facts for the active session.
 *
 * Only knowable after the handshake, so it lives beside the tunnel rather than in the
 * profile. It includes both server-pushed settings and capability-negotiated transport
 * modes. Two deliberate limits:
 *
 * * [routes] is CAPPED at [ROUTE_SAMPLE] entries with [routeCount] carrying the real total.
 *   A server may advertise an arbitrarily long list — an operator pushing a country-sized
 *   prefix set is a normal thing to do — and the detail sheet builds a view per row with no
 *   recycling. Capping at the source keeps both the snapshot and the sheet bounded instead
 *   of hoping the list stays short.
 * * the session token is NOT here and must never be: it is the credential that authorises a
 *   bonded stream to join this session.
 */
data class PushedFacts(
    val routes: List<String> = emptyList(),
    val routeCount: Int = 0,

    /**
     * How many of [routeCount] the builder actually took, or `-1` before the TUN exists.
     *
     * The card used to show [routeCount] — the number the server SENT — as though it were the
     * number in force. Those differ whenever a route is malformed or the builder rejects it:
     * `addRoute` throws, the failure is logged, and the tunnel comes up carrying less than the
     * card claims. That is the one direction a protection card must never be wrong in, so the
     * installed count is tracked separately and published only once `establish()` has returned.
     *
     * `-1` is deliberately not `0`: "not installed yet" and "installed none" look identical
     * otherwise, and the first is a normal moment during connect while the second is a fault.
     */
    val routesInstalled: Int = -1,
    val multipathAdaptive: Boolean = false,
    val paddingEnabled: Boolean = false,
    val paddingMin: Int = 0,
    val paddingMax: Int = 0,
    val heartbeatEnabled: Boolean = false,
    val heartbeatIntervalMs: Long = 0,
    val shapingEnabled: Boolean = false,
    val familyMode: String? = null,
    val carrierAddress: String? = null,
    val recordizerMode: String? = null,
    val recordizerPolicy: String? = null,
    val roamingMode: String? = null,
    val roamingPolicy: String? = null,
) {
    companion object {
        /** How many pushed routes the UI ever holds or renders. */
        const val ROUTE_SAMPLE = 6
    }
}

/** How much of the device's traffic the tunnel carries. */
enum class ProtectionScope {
    /** Every app, every route. */
    ALL,

    /** Only the apps the user picked (`apps_mode = include`). */
    ONLY_SELECTED,

    /** Every app except the ones the user picked (`apps_mode = exclude`). */
    ALL_EXCEPT,

    /** Split tunnel: only the configured/pushed routes go through the VPN. */
    SPLIT_ROUTES,
}

/** Something that narrows what the tunnel protects, worth telling the user about. */
enum class ProtectionWarning {
    /** `allow_lan` — RFC1918 is carved out, so LAN traffic is not in the tunnel. */
    LAN_OUTSIDE,

    /** `allow_ipv4_leak` — native IPv4 may bypass an IPv6-only full tunnel. */
    IPV4_OUTSIDE,

    /** `allow_ipv6_leak` — native IPv6 may bypass an IPv4-only full tunnel. */
    IPV6_OUTSIDE,

    /** Explicit `exclude` routes. */
    EXCLUDED_ROUTES,

    /** No pinned server key: the first connection trusts whoever answers (TOFU). */
    NO_PINNED_KEY,
}

/**
 * What a profile actually protects, derived from the profile alone.
 *
 * This backs a card that makes SECURITY CLAIMS, so the rule is: state only what the config
 * guarantees, and give anything that narrows the tunnel its own line rather than folding it
 * into a reassuring headline. A card that says "all traffic is protected" when it isn't is
 * worse than no card at all.
 *
 * Deliberately a pure function of [VpnConfig] with enum outputs — no Context, no string
 * resources — so the decisions are unit-testable and the wording stays localizable.
 * Runtime facts the profile cannot know (which resolver the server actually pushed, the
 * negotiated MTU, whether the system lockdown is on) are NOT guessed here; they arrive with
 * the tunnel snapshot.
 */
data class ProtectionSummary(
    val scope: ProtectionScope,
    /** Size of the per-app selection; meaningful for ONLY_SELECTED / ALL_EXCEPT. */
    val appCount: Int,
    val excludedRouteCount: Int,
    /** X25519 + ML-KEM-768. True for every wire mode except `plain`. */
    val postQuantum: Boolean,
    val dnsThroughTunnel: Boolean,
    val keyPinned: Boolean,
    val warnings: List<ProtectionWarning>,
) {
    /**
     * True only when nothing narrows what the tunnel carries — the one condition under
     * which the UI may claim "all traffic is protected".
     *
     * [ProtectionWarning.NO_PINNED_KEY] is excluded on purpose: pinning decides WHO the
     * client is willing to talk to, not HOW MUCH traffic it carries. It still gets its own
     * warning line.
     */
    val carriesEverything: Boolean
        get() = scope == ProtectionScope.ALL &&
            warnings.none { it != ProtectionWarning.NO_PINNED_KEY }

    companion object {
        /**
         * @param globalAllowLan the app-wide Settings toggle. It MUST be passed, because the
         * tunnel carves the private ranges out on `config.allowLan || globalAllowLan`
         * (QeliService) — reading only the profile field made the card claim "all traffic is
         * protected" while RFC1918, link-local and multicast went past the VPN. A card that
         * makes security claims has to be wrong in the SAFE direction, and this was wrong in
         * the other one. (Audit 2026-08-02, §6.)
         */
        @JvmOverloads
        fun of(config: VpnConfig, globalAllowLan: Boolean = false): ProtectionSummary {
            val apps = config.apps.size
            val scope = when {
                config.appsMode.equals("include", ignoreCase = true) -> ProtectionScope.ONLY_SELECTED
                config.appsMode.equals("exclude", ignoreCase = true) -> ProtectionScope.ALL_EXCEPT
                !config.isFullTunnel -> ProtectionScope.SPLIT_ROUTES
                else -> ProtectionScope.ALL
            }
            val warnings = buildList {
                // The compact strip can show only the first warning. Missing-family egress is
                // the broadest bypass, so keep both family escape hatches ahead of narrower
                // LAN/route exceptions.
                if (config.allowIpv4Leak) add(ProtectionWarning.IPV4_OUTSIDE)
                if (config.allowIpv6Leak) add(ProtectionWarning.IPV6_OUTSIDE)
                if (config.excludeRoutes.isNotEmpty()) add(ProtectionWarning.EXCLUDED_ROUTES)
                // allow_lan only carves routes out of a full-tunnel capture. In split mode it
                // must not imply that authenticated include/pushed routes were subtracted.
                if (config.isFullTunnel && (config.allowLan || globalAllowLan)) {
                    add(ProtectionWarning.LAN_OUTSIDE)
                }
                if (config.serverPublicKeyHex.isNullOrEmpty()) add(ProtectionWarning.NO_PINNED_KEY)
            }
            return ProtectionSummary(
                scope = scope,
                appCount = apps,
                excludedRouteCount = config.excludeRoutes.size,
                // Every mode runs the hybrid PQ ClientHello except `plain`, which uses a
                // raw X25519 exchange (QeliService: performHandshakePlain vs
                // performHandshake). obfs and reality-tls are transport wrappers around the
                // SAME PQ handshake, so they count as post-quantum.
                postQuantum = !config.wireMode.equals("plain", ignoreCase = true),
                // Explicit resolvers are reached through the tunnel; a full tunnel captures
                // DNS regardless. Anything narrower cannot be claimed from the profile
                // alone, so it is reported as system DNS until the snapshot says otherwise.
                dnsThroughTunnel = config.dnsServers.isNotEmpty() || config.isFullTunnel,
                keyPinned = !config.serverPublicKeyHex.isNullOrEmpty(),
                warnings = warnings,
            )
        }
    }
}

/**
 * Non-secret properties of the immutable configuration owned by the running VPN generation.
 *
 * The profile editor is intentionally usable while connected. UI code must therefore not
 * parse the currently saved profile when it describes the live connection: those edits do
 * not take effect until reconnect and may even refer to a different active profile. Keep the
 * password, session token and device identity out of this process-wide snapshot.
 */
data class LiveConnectionProperties(
    val serverAddress: String,
    val port: Int,
    val wireMode: String,
    val protocol: String,
    val quicEnabled: Boolean,
    val configuredMtu: Int,
    val reconnectEnabled: Boolean,
    val protection: ProtectionSummary,
) {
    val displayEndpoint: String
        get() = if (serverAddress.contains(':')) "[$serverAddress]:$port" else "$serverAddress:$port"

    companion object {
        fun of(config: VpnConfig, globalAllowLan: Boolean): LiveConnectionProperties =
            LiveConnectionProperties(
                serverAddress = config.serverAddress,
                port = config.port,
                wireMode = config.wireMode,
                protocol = config.protocol,
                quicEnabled = config.quicEnabled,
                configuredMtu = config.mtu,
                reconnectEnabled = config.reconnectEnabled,
                protection = ProtectionSummary.of(config, globalAllowLan),
            )
    }
}
